### DISCUSSION TITLE
Show "Open File" Dialog From Rust?

### DISCUSSION THREAD

**Discussion by swissmanu** [tauri-apps/tauri#3275] [Q&A] [answered]
I am diving into implementing my own menus as described here https://tauri.studio/docs/guides/menu

So far, I defined a `CustomMenuItem` labeled `Open`. My goal is to let the user select a file and then do some processing on that file. My first intuition was to emit a global `"showFileDialog"` event once the user clicked the `CustomMenuItem`. I'd then handle that event in the frontend and use the `dialog` API (https://tauri.studio/docs/api/js/modules/dialog) to show the file dialog.

Is this the way to go, or is there a more elegant way where I show the file dialog "directly" from within Rust-world, without a detour to JS-land first?

**Comment by xuchaoqian** [comment, marked as answer]
Use the Dialog APIs from rust: [The Dialog API module allows you to show messages and prompt for file paths.](https://docs.rs/tauri-api/latest/tauri_api/dialog/index.html)

**Reply by swissmanu** [reply]
This looks kind of too obvious... 🙈 ... Thank you! Will give it a try.

**Comment by swissmanu** [comment]
After RTFM, this is what I ended up with:

```rust
use tauri::api::dialog;
use tauri::{CustomMenuItem, Menu, MenuItem, Submenu};

fn main() {
  let open = CustomMenuItem::new("open".to_string(), "Open");
  let fileMenu = Submenu::new("File", Menu::new().add_item(open));
  let menu = Menu::new()
    .add_submenu(fileMenu)
    .add_native_item(MenuItem::Separator)
    .add_native_item(MenuItem::Quit);

  tauri::Builder::default()
    .menu(menu)
    .on_menu_event(|event| match event.menu_item_id() {
      "open" => {
        dialog::FileDialogBuilder::default()
          .add_filter("Markdown", &["md"])
          .pick_file(|path_buf| match path_buf {
            Some(p) => {}
            _ => {}
          });
      }
      _ => {}
    })
    .run(tauri::generate_context!())
    .expect("error while running tauri application");
}
```

Please bear with me... Still new to Rust 🦀😇

**Reply by Ivorforce** [reply]
To be honest, it's still kinda obscure what you need to do at this point, because you probably want a window reference inside there. Rust disallows that, of course, because all you're passed is an expirable reference.
Long story short, I cache the window ID and the app handle (which thankfully doesn't expire) to reference it later:
```rust
            let window = event.window();
            let window_name = window.label().to_string();
            let app = window.app_handle();
            
            match event.menu_item_id() {
              "openFile" => {
                dialog::FileDialogBuilder::default()
                    .pick_file(move |path_buf| match path_buf {
                        Some(p) => {
                            app.windows()[window_name.as_str()].emit_all("openFile", p).unwrap();
                        }
                        _ => {}
                    });

```

**Reply by JonBackupRepos** [reply]
Hi Ivorface, very helpful answer.  You wouldn't perhaps know how to receive that custom event in the frontend if it's using Yew?  I can handle message passing front to back, but not back to front somehow with Yew-Tauri builds.

**Reply by Ivorforce** [reply]
@JonBackupRepos Sorry, I don't know anything about yew. I don't know why it would make a difference though; tauri is managing the passed messages and you have to `listen` to it yourself on the frontend. [Here's how I do it for React](https://github.com/Ivorforce/Git-Diff-Stepper/blob/9f18a9a2e526c5d334f7553b796525ff0e2f89da/src/view/App.tsx#L16).

**Reply by JonBackupRepos** [reply]
@Ivorforce, sadly, I have to use WASM for this application.  Found a way to handle the listen, involves tauri_sys for API access and some futures work, but working now.  Thanks anyway

**Reply by Nikitf777** [reply]
Why when I'm trying to use `tauri::api::dialog`, it says "could not find `api` in `tauri`"?

**Reply by FabianLars** [reply]
because `tauri::api` is a v1 thing. In v2 most of the api modules were moved to separate plugins, in this case https://v2.tauri.app/plugin/dialog/
